ulog.cpp

```diff
@@ -0,0 +1,316 @@
+        log_error_errno(errno, "Unable to open ULog file(%s): (%m)", path);
+        return false;
+    }
+
+    _timeout = _timeout_manager->add_timeout(MSEC_PER_SEC, _ulog_timeout_cb, this);
+    null_check(_timeout, false);
+
+    _waiting_header = true;
+    _waiting_first_msg_offset = false;
+    _expected_seq = 0;
+    _buffer_len = 0;
+    _system_id = SYSTEM_ID;
+
+    log_info("Logging system_id=%u on %s", _system_id, path);
+    return true;
+}
+
+void ULog::stop()
+{
+    mavlink_message_t msg;
+    mavlink_command_long_t cmd;
+
+    if (!_file) {
+        log_error("ULog not started");
+        return;
+    }
+
+    bzero(&cmd, sizeof(cmd));
+    cmd.command = MAV_CMD_LOGGING_STOP;
+    cmd.target_component = MAV_COMP_ID_ALL;
+    cmd.target_system = _target_system_id;
+
+    mavlink_msg_command_long_encode(_system_id, MAV_COMP_ID_ALL, &msg, &cmd);
+    _send_msg(&msg, _target_system_id);
+
+    if (_timeout) {
+        _timeout_manager->del_timeout(_timeout);
+        _timeout = nullptr;
+    }
+
+    fflush(_file);
+    fclose(_file);
+    _file = NULL;
+    _system_id = 0;
+}
+
+int ULog::write_msg(const struct buffer *buffer)
+{
+    const bool mavlink2 = buffer->data[0] == MAVLINK_STX;
+    uint32_t msg_id;
+
+    _stat.write.total++;
+    _stat.write.bytes += buffer->len;
+
+    if (mavlink2) {
+        struct mavlink_router_mavlink2_header *msg
+            = (struct mavlink_router_mavlink2_header *)buffer->data;
+        msg_id = msg->msgid;
+    } else {
+        struct mavlink_router_mavlink1_header *msg
+            = (struct mavlink_router_mavlink1_header *)buffer->data;
+        msg_id = msg->msgid;
+    }
+
+    /* Check if we are interested in this msg_id */
+    if (msg_id != MAVLINK_MSG_ID_COMMAND_ACK && msg_id != MAVLINK_MSG_ID_LOGGING_DATA_ACKED
+        && msg_id != MAVLINK_MSG_ID_LOGGING_DATA) {
+        return buffer->len;
+    }
+
+    /*
+     * As MAVLINK_MSG_ID_LOGGING_DATA_ACKED and MAVLINK_MSG_ID_LOGGING_DATA
+     * are MAVLink 2 only messages lets untrim all messages that match
+     * with the interested msg_ids.
+     */
+    uint8_t data[MAVLINK_MAX_PACKET_LEN * 4];
+    struct buffer msg_buffer { 0, data };
+    uint8_t *payload;
+
+    untrim_msg(buffer, &msg_buffer);
+    buffer = &msg_buffer;
+
+    if (mavlink2)
+        payload = buffer->data + sizeof(struct mavlink_router_mavlink2_header);
+    else
+        payload = buffer->data + sizeof(struct mavlink_router_mavlink1_header);
+
+    /* Handle messages */
+    switch (msg_id) {
+    case MAVLINK_MSG_ID_COMMAND_ACK: {
+        mavlink_command_ack_t *cmd = (mavlink_command_ack_t *)payload;
+
+        if (!_timeout || cmd->command != MAV_CMD_LOGGING_START)
+            break;
+
+        if (cmd->result == MAV_RESULT_ACCEPTED) {
+            _timeout_manager->del_timeout(_timeout);
+            _timeout = NULL;
+        } else
+            log_error("MAV_CMD_LOGGING_START result(%u) is different than accepted", cmd->result);
+        break;
```

Comment:
Why not? when the result is MAV_RESULT_ACCEPTED it is zero so it will be trimmed.